### PR TITLE
#95 Added calculateMeans function

### PR DESCRIPTION
**Pull request by hannaei**

### CODE REVIEW

**Comment by IndrajeetPatil** [comment]
The build is failing because the git submodule directory (`\scripts`) is not present in this branch. 
I am guessing the `issue95` branch was created before commit that resolved #91 was merged into `develop`? 🤔

**Comment by IndrajeetPatil** [comment]
Actually, that folder seems to have been deleted in the PR. Was this done deliberately?!

image

**Comment by IndrajeetPatil** [comment]
@PavelBal Can you please have a look at this soon? 
I want to make some changes, but don't want to run into merge conflicts. 😅

**Comment by PavelBal** [comment]
The returned `DataSet` should have the name `Mean`.

**Comment by IndrajeetPatil** [comment]
Resolved merge conflict.

**Comment by PavelBal** [comment]
@IndrajeetPatil Can you check why the build is failing? Again something with enum, I don't understand why it does not work any more, as was supposed to work exactly as if it were a part of ospsuite?

**Review by PavelBal — CHANGES_REQUESTED** [review]

**Comment by PavelBal** [comment]
@IndrajeetPatil Take a look at calculateMeanDataSet(). If you don't have any coomments, I will merge.

**Review by PavelBal — APPROVED** [review]

**Comment by PavelBal** [comment]
@IndrajeetPatil Erm, why is that?:

```
Error : 'enum' is not an exported object from 'namespace:ospsuite'
Error: unable to load R code in package 'esqlabsR'
Execution halted
ERROR: lazy loading failed for package 'esqlabsR'
```

**Comment by IndrajeetPatil** [comment]
This has to be related to _ospsuite_ `dll`s being out of date or something. 
I am seeing similar failure in sensitivity analysis PR.

File: R/utilities-data.R [hunk 1]
```diff
@@ -133,3 +127,158 @@ stringToNum <- function(string) {
...
+  if (isEmpty(df)){
+    meanDataSet$addMetaData(name = "Subject ID", value = "mean")
+    return(meanDataSet)
+  }
+
+  molWeights <- unique(df$molWeight)
+  # molWeight is specified by user - use as specified
+  if (!is.null(outputMolWeight)) {
+    meanDataSet$molWeight <- outputMolWeight
+  } else
```
Comment (IndrajeetPatil) [inline R/utilities-data.R:175]: These are `else if` statements, but they are indented as if the `if` is nested inside `else`, which makes them really difficult to understand.

File: R/utilities-data.R [hunk 2]
```diff
@@ -133,3 +127,158 @@ stringToNum <- function(string) {
...
+    geometric = {
+      yMeans <- tapply(df[, "yValues"], df[, "xValues"], geomean)
+      yError <- tapply(df[, "yValues"], df[, "xValues"], geosd)
+      meanDataSet$setValues(xValues = as.numeric(names(yMeans)), yValues = yMeans, yErrorValues = yError)
+      meanDataSet$yErrorType <- ospsuite::DataErrorType$GeometricStdDev
+    }
+  )
+
+  # add all meta that are equal in every data set
+  metaDataNames <- Reduce(intersect, lapply(c(dataSets), function(x) {
```
Comment (IndrajeetPatil) [inline R/utilities-data.R:265]: Nice!

File: R/utilities-data.R [hunk 3]
```diff
@@ -133,3 +127,158 @@ stringToNum <- function(string) {
...
+#' @param outputMolWeight molWeight of output data set in `g/mol` - obligatory when initial
+#'  data sets have differing molWeight values
+#' @details Calculates mean and standard deviation of the yValues of the given `DataSet`
+#'  objects per xValue. The meta data of the returned `DataSet` consists of all meta
+#'  data that are equal in all initial data sets. Its LLOQ is the mean LLOQ value of all
+#'  data sets which have an LLOQ set, e.g. if dataSet1 has LLOQ 1, dataSet2 has LLOQ 3
+#'  and dataSet3 has no LLOQ, then 2 is used for the returned `DataSet`.
+#'  The LLOQ of the returned `DataSet` is the arithmetic mean of LLOQ values of
+#'  all `DataSet`s
+#' @return A single `DataSet` object
```
Comment (IndrajeetPatil) [inline R/utilities-data.R:153]: The parameter documentation says:

> #' @param dataSets list of `DataSet` objects

So if I understand this `@return` section correctly, the function takes in a list of `DataSet` objects and returns only a single `DataSet` object?! Can that be right?

**Review by IndrajeetPatil — CHANGES_REQUESTED** [review]

File: R/utilities-data.R (hunk 1 above)
Comment (PavelBal) [reply R/utilities-data.R]: Changed indentation.

File: R/utilities-data.R (hunk 3 above)
Comment (PavelBal) [reply R/utilities-data.R:153]: Yes, that's right.

**Review by IndrajeetPatil — APPROVED** [review]

**Pull request merged**